### PR TITLE
OneDeploy API - /publish and /api/publish

### PR DESCRIPTION
**Pull request by shrishrirang**
- Adds a new /publish and /api/publish APIs for Java deployment scenarios
- This new API primarily supports Java scenarios currently, but can be easily extended for additional scenarios
- Overview of new changes
  - OneDeploy reuses bulk of the code from PushDeploymentController and DeploymentManager
  - OneDeployBuilder is a new OneDeploy specific site builder, OneDeployFetch is a OneDeploy specific Fetch handler
  - WatchedFileEnabled/RestartAllowed/ArtifactType/TargetRootPath are additional knobs introduced in DeploymentInfoBase.cs for OneDeploy specific functionality
  - The implementation inspects the WEBSITE_STACK environment variable for certain scenarios. This variable is injected by the App Service runtime
  - OneDeploy uses the new restart API (in other words, it does not rely on touching any file for triggering a recycle)
- Summary of changes to existing code:
  - A bunch of variables were renamed for readability purposes
  - ZipDeploymentInfo is now ArtifactDeploymentInfo (generalized) and is now used by OneDeploy for non-zip artifacts as well
- Note: This is primarily a port of https://github.com/projectkudu/kudu/pull/3203 with some additional enhancements to OneDeploy. These enhancements will be back ported to https://github.com/projectkudu/kudu

### CODE REVIEW

**Comment by shrishrirang** [comment]
@jvano @sanchitmehta

**Comment by shrishrirang** [comment]
@jvano: Added optional query parameter, `"ignorestack=<true|false>"` to be able to override `WEBSITE_STACK`. Say when you are deploying app.war to a site that is not a Tomcat site. Example: a custom container or HttpPlatformHandler site. 

Kudu-windows uses `'stack=<tomcat|java|jbosseap>'` query parameter instead. That will be removed in favor of `ignorestack`

**Comment by jvano** [comment]
Sounds good!

:shipit:

**Comment by jvano** [comment]
:shipit:

**Comment by shrishrirang** [comment]
@sanchitmehta: Can you merge this? This is needed for ANT90.

File: Kudu.Contracts/Deployment/ArtifactType.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Kudu.Contracts.Deployment
+{
+    public enum ArtifactType
+    {
+        Invalid,
```
Comment (sanchitmehta) [inline Kudu.Contracts/Deployment/ArtifactType.cs:11]: What does the Invalid ArtifactType represent ? Is it that when we can't recognize the artifact type from the query param ? If yes, could we rename it to `Unknown`

File: Kudu.Core/Deployment/DeploymentManager.cs [hunk 2]
```diff
@@ -293,12 +294,24 @@ public void Delete(string id)
...
+        public async Task RestartMainSiteIfNeeded(ITracer tracer, ILogger logger, DeploymentInfoBase deploymentInfo)
         {
             // If post-deployment restart is disabled, do nothing.
             if (!_settings.RestartAppOnGitDeploy())
             {
                 return;
+            }
+
+            // Proceed only if 'restart' is allowed for this deployment
+            if (deploymentInfo != null && !deploymentInfo.RestartAllowed)
```
Comment (sanchitmehta) [inline Kudu.Core/Deployment/DeploymentManager.cs:306]: Is `RestartAllowed` added as a OneDeploy only property ?

**Review by sanchitmehta — APPROVED** [review]

File: Kudu.Contracts/Deployment/ArtifactType.cs (hunk 1 above)
Comment (shrishrirang) [reply Kudu.Contracts/Deployment/ArtifactType.cs]: Done

File: Kudu.Core/Deployment/DeploymentManager.cs (hunk 2 above)
Comment (shrishrirang) [reply Kudu.Core/Deployment/DeploymentManager.cs:306]: No. But OneDeploy is the only scenario that changes the default. For everyone else, `RestartAllowed=true` (default behavior)

**Comment by shrishrirang** [comment]
@sanchitmehta : I don't have access to the repo, so you'll have to merge it for me whenever it looks good to you :-)

**Comment by sanchitmehta** [comment]
Done. Merged and given you the access

**Comment by shrishrirang** [comment]
> Done. Merged and given you the access

Thanks @sanchitmehta !

**Pull request merged**